### ISSUE TITLE
string trouble after JSON parsing

### ISSUE THREAD

**Issue by jdall** [dvidelabs/flatcc#68]
Having parsed a JSON string with success, I now try to extract the details from the flatbuffer. Below is the output from the app. First the JSON string is dumped, then the flatbuffer, and finally info on extracting the details
```
{
  "msg_type": "Ping",
  "msg": {
    "str": "123"
  }
}

successfully parsed 7 lines
parsed content:
  0000  08 00 00 00 50 49 50 4f  de ff ff ff 08 00 00 00  ....PIPO........
  0010  01 00 00 00 f0 ff ff ff  04 00 00 00 03 00 00 00  ................
  0020  31 32 33 00 06 00 08 00  04 00 08 00 09 00 08 00  123.............
  0030  04 00                                            ..
input size: 58, output size: 50
const unsigned char pingpong[] = {
    0x08, 0x00, 0x00, 0x00, 0x50, 0x49, 0x50, 0x4f,    0xde, 0xff, 0xff, 0xff, 0x08, 0x00, 0x00, 0x00,
    0x01, 0x00, 0x00, 0x00, 0xf0, 0xff, 0xff, 0xff,    0x04, 0x00, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00,
    0x31, 0x32, 0x33, 0x00, 0x06, 0x00, 0x08, 0x00,    0x04, 0x00, 0x08, 0x00, 0x09, 0x00, 0x08, 0x00,
    0x04, 0x00,
};
Ping message detected
Ping.str:
  0000  ff ff ff 04 00 00 00 03  00 00 00 31 32 33 00 06  ...........123..
Ping.str=4026531840:"���"
``` 
As can be seen, I manage to detect the Ping message, and apparently I also get the string (at least I see the "123" text in the hex dump) - but the string length returned by nsc(string_len(s)) is way off (4026531840 instead of 3).

I've been reading the docs and the flatcc code with a magnifying glass, but I just can't identify what I am doing wrong. The big difference from my code and the test code, seems to be that I'm accessing a string in a table in a union.
What have I missed? (using latest flatcc master branch - december 20)

The fbs contains:
```
namespace eurisco.modiccp.mockup.pingpong;
attribute "priority";
union Message { Ping, Pong }
table Ping {
  str:string;
}
table Pong {
  str:string;
}
table Body {
  msg:Message;
}
root_type Body;
file_identifier "PIPO";
file_extension "msg";
```
And the test code is as follows:
```
#include <pingpong_builder.h>
#include <pingpong_reader.h>
#include <pingpong_verifier.h>
#include <pingpong_json_parser.h>
#include <pingpong_json_printer.h>
#include "flatcc/support/hexdump.h"
#include "flatcc/support/cdump.h"
#include "flatcc/support/readfile.h"

#ifdef NDEBUG
#define COMPILE_TYPE "(optimized)"
#else
#define COMPILE_TYPE "(debug)"
#endif

#define FILE_SIZE_MAX (1024 * 10)

#undef ns
#define ns(x) FLATBUFFERS_WRAP_NAMESPACE(eurisco_modiccp_mockup_pingpong, x)
#undef nsc
#define nsc(x) FLATBUFFERS_WRAP_NAMESPACE(flatbuffers, x)

int verify_parse (void *buffer) {
  ns(Body_table_t) body = ns(Body_as_root_with_identifier)
      (buffer, ns(Body_identifier));
  ns(Ping_table_t) ping = ns(Ping_as_root_with_identifier)
      (buffer, ns(Ping_identifier));
  ns(Pong_table_t) pong = ns(Pong_as_root_with_identifier)
      (buffer, ns(Pong_identifier));
  ns(Message_union_type_t) msgType;
  nsc(string_t) pingStr;
  
  if (!(body = ns(Body_as_root(buffer)))) {
    printf("Body not available\n");
    return -1;
  }

  msgType = ns(Body_msg_type(body));
  switch (msgType) {
  case ns(Message_Ping):
    printf("Ping message detected\n");
    if (!(ping = ns(Ping_as_root(buffer)))) {
      printf("Ping not available\n");
      return -1;
    }
    if (!(pingStr = ns(Ping_str(ping)))) {
      printf("Ping.str not available\n");
      return -1;
    }  
    hexdump("Ping.str", (void*)pingStr, 16, stdout);
    printf("Ping.str=%zd:\"%s\"\n", nsc(string_len(pingStr)), (char*)pingStr);
    break;
  case ns(Message_Pong):
    printf("Pong message detected\n");
    break;
  default:
    printf("Unexpected message detected\n");
    break;
  }
  return 0;
}

void testParseJsonBuffer (const char *buf, size_t in_size) {
  void *flatbuffer = NULL;
  size_t out_size;
  flatcc_json_parser_t ctx;
  flatcc_builder_t builder, *B = &builder;
  int ret = -1, flags = 0;

  printf("%s\n", buf);
  flatcc_builder_init(B);
  if (!pingpong_parse_json(B, &ctx, buf, in_size, flags)) {
    printf("successfully parsed %d lines\n", ctx.line);
    flatbuffer = flatcc_builder_finalize_aligned_buffer(B, &out_size);

    hexdump("parsed content", flatbuffer, out_size, stdout);
    printf("input size: %lu, output size: %lu\n",
        (unsigned long)in_size, (unsigned long)out_size);

    cdump("pingpong", flatbuffer, out_size, stdout);
    verify_parse(flatbuffer);
  } else {
    printf("%d:%d: %d %s\n",
        (int)ctx.line, (int)(ctx.error_loc - ctx.line_start + 1),
        ctx.error, flatcc_json_parser_error_string(ctx.error));
  }

  if (flatbuffer) {
    flatcc_builder_aligned_free(flatbuffer);
  }
  flatcc_builder_clear(B);
}

void testParseJsonFile (const char *filename) {
  char *buf = NULL;
  size_t in_size;
  buf = readfile(filename, FILE_SIZE_MAX, &in_size);
  if (buf) {
    testParseJsonBuffer(buf, in_size);
    free((void *)buf);
  } else {
    printf("%s: could not read input json file\n", filename);
  }
}
```

**Comment by mikkelfj** [comment]
Hi jdall, I'll follow up on this. A very quick answer: the string is encoded correctly in the originally parsed buffer:

The length is ` 03 00 00 00` and is aligned to a 4 byte boundary. This is followed by the text `31 32 33 ` followed by a terminal `00`.

If you subtract ((char *)s-(char *)buffer_start) you should get 32 because that is where the string starts if you have the correct string pointer, which you likely don't.

Digging further at appears that use Ping_as_root as in `if (!(ping = ns(Ping_as_root(buffer)))) {` but root is only for the root table, not other tables. There should be some union examples in builder.md, the tutorial, or in json_test.c. `Body` is the root table and you do get that right initially, it seems: `if (!(body = ns(Body_as_root(buffer)))) {`

It's late and I'm half taken by the flu, so you need to dig a bit more yourself. But I'd be happy to answer more questions later.

**Comment by mikkelfj** [comment]
https://github.com/dvidelabs/flatcc/blob/master/samples/monster/monster.c#L81

**Comment by mikkelfj** [comment]
The previous one as construction, this is read access for a union
https://github.com/dvidelabs/flatcc/blob/master/samples/monster/monster.c#L283

**Comment by mikkelfj** [comment]
This should do the trick. You problem is that you try to assign Ping or Pong as  buffer roots, but a buffer can only have one root, here Body. So the union member must be accessed via ns(Body_msg)(body) assigned to the proper ns(Ping_table_t) or ns(Pong_table_t) depending on ns(Body_msg_type)(body).

Note that you can skip having a union wrapper by using type hashes where the buffer identifier is different and you use this to switch upon instead. It was intended for use cases where one of several message types are transmitted and it will generate slightly more compact buffer, but it doesn't work well with JSON because then you can't set the _type field in JSON and you can't have one JSON parser to handle all message types.

Fixed in source below, also with a some added return codes as well:

(the string hexdump is unsafe because the length might overflow)

```
#include <pingpong_builder.h>
#include <pingpong_reader.h>
#include <pingpong_verifier.h>
#include <pingpong_json_parser.h>
#include <pingpong_json_printer.h>
#include "flatcc/support/hexdump.h"
#include "flatcc/support/cdump.h"
#include "flatcc/support/readfile.h"

#ifdef NDEBUG
#define COMPILE_TYPE "(optimized)"
#else
#define COMPILE_TYPE "(debug)"
#endif

#define FILE_SIZE_MAX (1024 * 10)

#undef ns
#define ns(x) FLATBUFFERS_WRAP_NAMESPACE(eurisco_modiccp_mockup_pingpong, x)
#undef nsc
#define nsc(x) FLATBUFFERS_WRAP_NAMESPACE(flatbuffers, x)

int verify_parse (void *buffer) {
  ns(Body_table_t) body;

  ns(Ping_table_t) ping;
  ns(Pong_table_t) pong;
  ns(Message_union_type_t) msgType;
  nsc(string_t) pingStr;
  
  body = ns(Body_as_root_with_identifier)(buffer, ns(Body_identifier));
  if (!(body)) {
    printf("Body not available\n");
    return -1;
  }

  msgType = ns(Body_msg_type(body));
  switch (msgType) {
  case ns(Message_Ping):
    printf("Ping message detected\n");
    if (!(ping = ns(Body_msg(body)))) {
      printf("Ping not available\n");
      return -1;
    }
    if (!(pingStr = ns(Ping_str(ping)))) {
      printf("Ping.str not available\n");
      return -1;
    }  
    hexdump("Ping.str", (void*)pingStr, 16, stdout);
    printf("Ping.str=%zd:\"%s\"\n", nsc(string_len(pingStr)), (char*)pingStr);
    break;
  case ns(Message_Pong):
    printf("Pong message detected\n");
    break;
  default:
    printf("Unexpected message detected\n");
    break;
  }
  return 0;
}

int testParseJsonBuffer (const char *buf, size_t in_size) {
  void *flatbuffer = NULL;
  size_t out_size;
  flatcc_json_parser_t ctx;
  flatcc_builder_t builder, *B = &builder;
  int ret = -1, flags = 0;

  printf("%s\n", buf);
  flatcc_builder_init(B);
  if (!pingpong_parse_json(B, &ctx, buf, in_size, flags)) {
    printf("successfully parsed %d lines\n", ctx.line);
    flatbuffer = flatcc_builder_finalize_aligned_buffer(B, &out_size);

    hexdump("parsed content", flatbuffer, out_size, stdout);
    printf("input size: %lu, output size: %lu\n",
        (unsigned long)in_size, (unsigned long)out_size);

    cdump("pingpong", flatbuffer, out_size, stdout);
    ret = verify_parse(flatbuffer);
  } else {
    printf("%d:%d: %d %s\n",
        (int)ctx.line, (int)(ctx.error_loc - ctx.line_start + 1),
        ctx.error, flatcc_json_parser_error_string(ctx.error));
    ret -1;
  }

  if (flatbuffer) {
    flatcc_builder_aligned_free(flatbuffer);
  }
  flatcc_builder_clear(B);
  return ret;
}

int testParseJsonFile (const char *filename) {
  int ret;
  char *buf = NULL;
  size_t in_size;
  buf = readfile(filename, FILE_SIZE_MAX, &in_size);
  if (buf) {
    ret = testParseJsonBuffer(buf, in_size);
    free((void *)buf);
    return ret;
  } else {
    printf("%s: could not read input json file\n", filename);
    return -1;
  }
}

int main()
{
    return testParseJsonFile("pingpong.json");
}
```

Output is:

```
{
  "msg_type": "Ping",
  "msg": {
    "str": "123"
  }
}


successfully parsed 8 lines
parsed content:
  0000  08 00 00 00 50 49 50 4f  de ff ff ff 08 00 00 00  ....PIPO........
  0010  01 00 00 00 f0 ff ff ff  04 00 00 00 03 00 00 00  ................
  0020  31 32 33 00 06 00 08 00  04 00 08 00 09 00 08 00  123.............
  0030  04 00                                            ..
input size: 59, output size: 50
const unsigned char pingpong[] = {
    0x08, 0x00, 0x00, 0x00, 0x50, 0x49, 0x50, 0x4f,    0xde, 0xff, 0xff, 0xff, 0x08, 0x00, 0x00, 0x00,
    0x01, 0x00, 0x00, 0x00, 0xf0, 0xff, 0xff, 0xff,    0x04, 0x00, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00,
    0x31, 0x32, 0x33, 0x00, 0x06, 0x00, 0x08, 0x00,    0x04, 0x00, 0x08, 0x00, 0x09, 0x00, 0x08, 0x00,
    0x04, 0x00,
};
Ping message detected
Ping.str:
  0000  31 32 33 00 06 00 08 00  04 00 08 00 09 00 08 00  123.............
Ping.str=3:"123"
```

**Comment by mikkelfj** [comment]
One more issue to be aware of:

Due to FlatBuffers forward schema support, a future version of the schema may add more types to the union. Therefore the default case ` printf("Unexpected message detected\n");` is not necessarily unexpected, just unhandled, depending on your point of view.

FlatBuffers also allow for the union type NONE (= 0) when there is no message.

**Comment by jdall** [comment]
Thanks, Mikkel.

It's very nice of you to let me know in such a detailed way where I went wrong. 
And thank you for all the hints - being a flatbuffers noob, that's quite valuable.

Do you by any chance know when v0.5.0 is expected to be released?

Thanks again for all your help - it's truly great support.

And best wishes for the season holidays - hopefully flu-less.

**Comment by mikkelfj** [comment]
I'm glad you find it useful and happy holidays to you too.

It's hard to give a date on 0.5.0 because it is a complex feature release with ongoing work where 0.4.3 was  a bug-fix release. It depends on how much time I can put into it going forward, but I am working on it with priority and hoping to release within 1-2 months. There has been some breaking interface changes on master, and more might come, so now is not a good time to do a release.

I'd also like to give users a chance to try out the new features before releasing. If you want to test union vectors in JSON, or base64 encoded uint8 vectors in JSON, please don't hesitate to ask, it would be helpful.

If you need a stable release with the bug affecting you and the deallocation feature, I recommend creating you branch backporting those. The only other bug fix on 0.5.0 is a regression in the verifier that is not very likely to hit, so there isn't really enough material to do an official 0.4.4 bug-fix release.

**Comment by jdall** [comment]
Thanks for the update, Mikkel.

I'll stay with the patched 0.4.3 branch that I've already created, unless I really need some features only supported in the 0.5 branch. 

Please reconsider making a 0.4.4 with at least the 1cb2664 and 1cb2664 patches - taking into account the time I spent on debugging this, I think it's relevant. 
That said, I fully understand if you are hesitant to do it. If the user base is small, it may not really be worth your time.

**Comment by mikkelfj** [comment]
I updated the v0.4.3 note in the README: https://github.com/dvidelabs/flatcc/commit/53f4f3e7b39d748480a46110074c3e2b4407d5d0

I don't think the JSON bug is very common, I only saw it a month ago, it depends on the names in a given schema, and not all users use JSON. Doing a release also affects a few external packages so I don't want to release for every single bug-fix.

**Comment by mikkelfj** [comment]
@jdall : If you plan on switching to 0.5.0, now would be a good time to test it out prior to the upcoming release.

I just added a fairly large update to master. Wrt. JSON this affects how JSON parses unions and generally adds support for unions that contain a mix of strings, structs and tables, and base64 encoding which was added a bit earlier. This brings the flatcc binary format on par with recent changes in the C++ version. I hope this covers the major breaking changes necessary prior to releasing 0.5.0 and I thus hope to be able to release soon.

I will be running some additional internal studies to verify the api stability going forward but I do not expect many new changes before a 0.5.0 release.
